### PR TITLE
IPython engine for knitr

### PR DESCRIPTION
**Pull request by mwouts**
Hello Yihui,

This is a proposal for an ipython engine in knitr.

Additional functionality with respect to the former python engine is;
* python objects persists from one chunk to the other
* image output is supported
* inline python code is supported

An example is available at https://github.com/mwouts/knitr/blob/master/inst/examples/knitr-ipython.Rmd

This contribution may require a fix in roxygen (replace 'roxygen_devel' with 'roxygen2', cf. https://github.com/mwouts/roxygen)

Also, I would like your feedback on the questions below:
* should the engine replace the 'python' one?
* should we expose (and document) the ipython class?

Regards,

Marc

### CODE REVIEW

**Comment by tmastny** [comment]
Any updates on this? I think it is a great idea

**Comment by mwouts** [comment]
Thanks tmastny. I am waiting for feedback on this from Yihui. 
Still, in the current state the fork works rather well, I use it frequently with complex python codes.
I do not know why the tests are not passing, but I would be available to work on fixing them if I'm told more about the issue.

**Comment by yihui** [comment]
Sorry for the late response, but this PR is too big for me to review (due to my very limited knowledge of Python and also limited time). I think the ideal way to move forward is to factor it out to a separate package, so that I don't have to maintain it in **knitr**, and you still have full control over this feature. If you need me to export certain functions so that you can use them in your package, just let me know. I hope we can hold on regarding `` `py ` `` and `` `ipython ` `` and think about it in the second step.

Basically this is how the recent new `python` engine was developed (in a separate package [**reticulate**](/rstudio/reticulate)); see #1440.

**Comment by mwouts** [comment]
Hello Yihui, yes I agree, providing the engine in a separate package would be a nice way to provide the functionality with less impact on `knitr`.

By the way, there is a significant overlap between my contribution (execute python code + retrieve plots and javascript) and the new `reticulate` package (execute python code + exchange data), so I will first contact the authors of `reticulate` to know more about there plans for supporting graphical output.

**Comment by yihui** [comment]
AFAIK, matplotlib plots are supported; see https://github.com/rstudio/reticulate/pull/107

You may open an issue in the [reticulate repo](https://github.com/rstudio/reticulate/issues) and discuss with them there.

**Comment by yihui** [comment]
Again, I'm not a Python expert, but I feel the reticulate package has done a reasonably good job: https://blog.rstudio.com/2018/03/26/reticulate-r-interface-to-python/ so I'm going to close this PR. If you want to develop this into a separate package and need my support from knitr, I'll be happy to. Many thanks for your contribution anyway!

**Comment by mwouts** [comment]
Hello Yihui, Tmastny,

Thanks for your feedback, you're right, this contribution would better fit as a standalone package.
I have repackaged the code as a separate package at https://github.com/mwouts/ipython_from_R, and it works well with knitr for building PDF presentations from latex + python files.

Now I'd like to move forward to interactive slides based on your very nice xaringan package. Tmastny, do you have an idea if this is already supported by reticulate ? With the ipython engine I do get the javascript plot, just for now it do not know how to display it in the slides (a more specific question for Yihui is as https://github.com/yihui/xaringan/issues/132)

Regards,

Marc
